node/lib/util/log_util.js

```diff
@@ -0,0 +1,165 @@
+ */
+exports.findMetaCommit = co.wrap(function *(repo,
+                                            metaCommit,
+                                            subCommit,
+                                            submoduleName,
+                                            subRepo,
+                                            subFetcher) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.instanceOf(metaCommit, NodeGit.Commit);
+    assert.instanceOf(subCommit, NodeGit.Commit);
+    assert.isString(submoduleName);
+    assert.instanceOf(subRepo, NodeGit.Repository);
+    assert.instanceOf(subFetcher, SubmoduleFetcher);
+
+    // First, resolve the committishes and throw if they cannot be resolved.
+
+    const subSha = subCommit.id().tostrS();
+
+    // While a recursive solution would be very simple, I'm afraid we'd
+    // overflow (v8 does not support tail call optimization at the moment).
+    // Furthermore, we need to do a breadth-first-search: we don't want to
+    // search the entire left side of a merge commit before checking the right.
+    //
+    // Strategy: walk back, starting from `metaCommit`, return the first
+    // commit where the sha for our submodule references (directly or
+    // indirectly) `subCommit`, but its parents don't.
+
+    let   toCheck = [metaCommit];   // commits left to check
+    const checked = new Set();      // SHAs checked
+    const existsInSha = new Map();  // repo sha to bool if target included
+
+    const doesExistInCommit = co.wrap(function *(commit) {
+        const sha = commit.id().tostrS();
+        if (existsInSha.has(sha)) {
+            return existsInSha.get(sha);
+        }
+        let result;
+
+        // Check to see if the repo references `subSha` directly, or one of
+        // its descendants.
+
+        const tree = yield commit.getTree();
+        let subShaForCommit;
+        try {
+            subShaForCommit = (yield tree.entryByPath(submoduleName)).sha();
+        }
+        catch (e) {
+            // Submodule doesn't exist in this commit; stop checking earlier
+            // commits.
+
+            result = null;
+        }
+        if (undefined !== subShaForCommit) {
+            if (subShaForCommit === subSha) {
+                result = true;
+            }
+            else {
+                // Ensure that the commit we're checking against is present.
+
+                yield subFetcher.fetchSha(subRepo,
+                                          submoduleName,
+                                          subShaForCommit);
+                result = (yield NodeGit.Graph.descendantOf(
+                                       subRepo,
+                                       NodeGit.Oid.fromString(subShaForCommit),
+                                       subCommit.id())) !== 0;
+            }
+        }
+        existsInSha.set(sha, result);
+        return result;
+    });
+
+    while (0 !== toCheck.length) {
+        const commit = toCheck.shift();
```

Comment:
The deque I linked to is backed by an array, so the perf should be quite good.  And in a very-branching repo, your queue size can get large.